pkg/apis/pipeline/v1/pipeline_types_test.go

```diff
@@ -129,6 +286,78 @@ func TestPipelineTask_ValidateRefOrSpec(t *testing.T) {
 	}
 }
 
+// TestPipelineTask_ValidateRefOrSpec_APIVersionsCompatibility validates `pipelineRef` and `pipelineSpec`
+// are guarded behind the appropriate feature flags i.e. alpha for now.
+func TestPipelineTask_ValidateRefOrSpec_APIVersionsCompatibility(t *testing.T) {
+	tests := []struct {
+		name    string
+		pt      PipelineTask
+		wc      func(ctx context.Context) context.Context
+		wantErr *apis.FieldError
+	}{
+		{
+			name: "pipelineRef requires alpha version",
+			pt: PipelineTask{
+				PipelineRef: &PipelineRef{},
+			},
+			wc: cfgtesting.EnableAlphaAPIFields,
+		}, {
+			name: "pipelineSpec requires alpha version",
+			pt: PipelineTask{
+				PipelineSpec: &PipelineSpec{},
+			},
+			wc: cfgtesting.EnableAlphaAPIFields,
+		}, {
+			name: "pipelineRef not allowed with beta version",
+			pt: PipelineTask{
+				PipelineRef: &PipelineRef{},
+			},
+			wc:      cfgtesting.EnableBetaAPIFields,
+			wantErr: apis.ErrGeneric("pipelineRef requires \"enable-api-fields\" feature gate to be \"alpha\" but it is \"beta\""),
+		}, {
+			name: "pipelineSpec not allowed with beta version",
+			pt: PipelineTask{
+				PipelineSpec: &PipelineSpec{},
+			},
+			wc:      cfgtesting.EnableBetaAPIFields,
+			wantErr: apis.ErrGeneric("pipelineSpec requires \"enable-api-fields\" feature gate to be \"alpha\" but it is \"beta\""),
+		}, {
+			name: "pipelineRef not allowed with stable version",
+			pt: PipelineTask{
+				PipelineRef: &PipelineRef{},
+			},
+			wc:      cfgtesting.EnableStableAPIFields,
+			wantErr: apis.ErrGeneric("pipelineRef requires \"enable-api-fields\" feature gate to be \"alpha\" but it is \"stable\""),
+		}, {
+			name: "pipelineSpec not allowed with beta version",
+			pt: PipelineTask{
+				PipelineSpec: &PipelineSpec{},
+			},
+			wc:      cfgtesting.EnableStableAPIFields,
+			wantErr: apis.ErrGeneric("pipelineSpec requires \"enable-api-fields\" feature gate to be \"alpha\" but it is \"stable\""),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx := context.Background()
+			if test.wc != nil {
+				ctx = test.wc(ctx)
+			}
+			err := test.pt.validateRefOrSpec(ctx)
+			if test.wantErr != nil {
+				if d := cmp.Diff(test.wantErr.Error(), err.Error()); d != "" {
+					t.Error(diff.PrintWantGot(d))
+				}
+			} else {
+				if err := test.pt.validateRefOrSpec(ctx); err != nil {
```

Comment:
nit: can this just be `if err != nil`, since the function is called above on line 347?

Reply:
that's a great catch, thanks @lbernick, done!